src/GafferScene/CollectPrimitiveVariables.cpp

```diff
@@ -0,0 +1,259 @@
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "IECoreScene/Primitive.h"
+
+#include "GafferScene/CollectPrimitiveVariables.h"
+
+#include "Gaffer/ArrayPlug.h"
+#include "Gaffer/Context.h"
+#include "Gaffer/StringPlug.h"
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+
+//////////////////////////////////////////////////////////////////////////
+// Internal utilities
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+static InternedString g_collectPrimitiveVariablesSuffix( "collectPrimitiveVariables:suffix" );
+
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// CollectPrimitiveVariables
+//////////////////////////////////////////////////////////////////////////
+
+IE_CORE_DEFINERUNTIMETYPED( CollectPrimitiveVariables );
+
+size_t CollectPrimitiveVariables::g_firstPlugIndex = 0;
+
+CollectPrimitiveVariables::CollectPrimitiveVariables( const std::string &name )
+	:	SceneElementProcessor( name, IECore::PathMatcher::NoMatch )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new StringVectorDataPlug( "suffixes", Plug::In, new StringVectorData() ) );
+	addChild( new StringPlug( "targetVariables", Plug::In, "P" ) );
+	addChild( new BoolPlug( "isVariation", Plug::Out, false ) );
+
+	// Fast pass-throughs for things we don't modify
+	outPlug()->attributesPlug()->setInput( inPlug()->attributesPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+	outPlug()->boundPlug()->setInput( inPlug()->boundPlug() );
+}
+
+CollectPrimitiveVariables::~CollectPrimitiveVariables()
+{
+}
+
+Gaffer::StringVectorDataPlug *CollectPrimitiveVariables::suffixesPlug()
+{
+	return getChild<StringVectorDataPlug>( g_firstPlugIndex + 0 );
+}
+
+const Gaffer::StringVectorDataPlug *CollectPrimitiveVariables::suffixesPlug() const
+{
+	return getChild<StringVectorDataPlug>( g_firstPlugIndex + 0 );
+}
+
+Gaffer::StringPlug *CollectPrimitiveVariables::targetVariablesPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *CollectPrimitiveVariables::targetVariablesPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+Gaffer::BoolPlug *CollectPrimitiveVariables::isVariationPlug()
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+const Gaffer::BoolPlug *CollectPrimitiveVariables::isVariationPlug() const
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+void CollectPrimitiveVariables::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneElementProcessor::affects( input, outputs );
+
+	if(
+		input == inPlug()->objectPlug() ||
+		input == suffixesPlug() ||
+		input == targetVariablesPlug()
```

Comment:
Looking below, it appears that `targetVariablesPlug()` doesn't affect `isVariationPlug()`. These days I'm tending to avoid dirtying multiple plugs in a single block, so it's easier to match the logic in the hash/compute methods. Longer term I'm thinking of having `affects<ThePlug>()` methods to enforce this separation of logic...